tests/wpt/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+fn create_context(wpt_path: &Path) -> (Context, logger::RecordingLogger) {
+    let mut context = Context::default();
+    let logger = logger::RecordingLogger::new();
+    boa_runtime::register(
+        &mut context,
+        RegisterOptions::new().with_console_logger(logger.clone()),
+    )
+    .expect("Failed to register boa_runtime");
+
+    // Define self as the globalThis.
+    let global_this = context.global_object();
+    context
+        .register_global_property(js_str!("self"), global_this, Attribute::all())
+        .unwrap();
+
+    // Define location to be an empty URL.
+    let location = Url::new("about:blank".to_string().into(), None.into())
+        .expect("Could not parse the location URL");
+    let location =
+        Url::from_data(location, &mut context).expect("Could not create the location URL");
+    context
+        .register_global_property(js_str!("location"), location, Attribute::all())
+        .unwrap();
+
+    let harness_path = wpt_path.join("resources/testharness.js");
+    let harness = Source::from_filepath(&harness_path).expect("Could not create source.");
+
+    if let Err(e) = context.eval(harness) {
+        panic!("Failed to eval testharness.js: {e:#?}");
+    }
+
+    (context, logger)
+}
+
+/// The result callback for the WPT test.
+fn result_callback__(
+    ContextData(logger): ContextData<logger::RecordingLogger>,
+    test: Test,
+    context: &mut Context,
+) -> JsResult<()> {
+    // Check the logs if the test succeeded.
+    assert_eq!(
+        test.status,
+        TestStatus::Pass,
+        "Test {:?} failed with message:\n  {:?}",
+        test.name.to_std_string_lossy(),
+        test.message.unwrap_or_default()
+    );
+
+    // Check the logs.
+    let logs = logger.all_logs();
+    if let Some(log_regex) = test.properties.get(&js_string!("logs")) {
+        if let Ok(logs_re) = log_regex.try_js_into::<Vec<JsValue>>(context) {
+            for re in logs_re {
+                let passes = if let Some(re) = re.as_regexp() {
+                    logs.iter().any(|log: &RecordingLogEvent| -> bool {
+                        let s = JsString::from(log.msg.clone());
+                        re.test(s, context).unwrap_or(false)
+                    })
+                } else {
+                    let re_str = re.to_string(context)?.to_std_string_escaped();
+                    logs.iter()
+                        .any(|log: &RecordingLogEvent| -> bool { log.msg.contains(&re_str) })
+                };
+                assert!(
+                    passes,
+                    "Test {:?} failed to find log: {}",
+                    test.name.to_std_string_lossy(),
+                    re.display()
+                );
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn complete_callback__(ContextData(test_done): ContextData<TestCompletion>) {
+    test_done.done();
+}
+
+#[derive(Debug, Clone, Trace, Finalize, JsData)]
+struct TestCompletion(Gc<AtomicBool>);
```

Comment:
This seems... wrong. `Gc` is thread-unsafe, and `AtomicBool` is used to add thread safety to things, so something's not working correctly. This should either be `Arc<AtomicBool>` if this is shared between multiple threads, or `Rc<Cell<bool>>` if this is not shared between threads.